src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java

```diff
@@ -61,32 +93,157 @@ public boolean isEmpty()
+     */
+    public Iterator<Pair<ByteComparable, Iterator<PrimaryKey>>> iterator(DecoratedKey min, DecoratedKey max)
     {
-        return index.iterator();
+        int minSubrange = min == null ? 0 : boundaries.getShardForKey(min);
+        int maxSubrange = max == null ? rangeIndexes.length - 1 : boundaries.getShardForKey(max);
+
+        List<Iterator<Pair<ByteComparable, PrimaryKeys>>> rangeLists = new ArrayList<>(maxSubrange - minSubrange + 1);
+        for (int i = minSubrange; i <= maxSubrange; i++)
+            rangeLists.add(rangeIndexes[i].iterator());
+
+        return MergeIterator.get(rangeLists, (o1, o2) -> ByteComparable.compare(o1.left, o2.left, ByteComparable.Version.OSS41),
+                                 new PrimaryKeysMergeReducer(rangeIndexes.length, clusteringComparator));
+    }
+
+    private static class PrimaryKeysMergeReducer extends MergeIterator.Reducer<Pair<ByteComparable, PrimaryKeys>, Pair<ByteComparable, Iterator<PrimaryKey>>>
+    {
+        private final Pair<ByteComparable, PrimaryKeys>[] toMerge;
+        private final int size;
+        private final Comparator<PrimaryKey> comparator;
+
+        @SuppressWarnings("unchecked")
+        PrimaryKeysMergeReducer(int size, ClusteringComparator clusteringComparator)
+        {
+            this.toMerge = new Pair[size];
+            this.size = size;
+            this.comparator = (o1, o2) -> {
+                if (!o1.partitionKey().equals(o2.partitionKey()))
+                    return o1.partitionKey().compareTo(o2.partitionKey());
+                return clusteringComparator.compare(o1.clustering(), o2.clustering());
+            };
+        }
+
+        @Override
+        public void reduce(int idx, Pair<ByteComparable, PrimaryKeys> current)
+        {
+            toMerge[idx] = current;
+        }
+
+        @Override
+        public Pair<ByteComparable, Iterator<PrimaryKey>> getReduced()
+        {
+            ByteComparable term = getTerm();
+
+            List<Iterator<PrimaryKey>> keyIterators = new ArrayList<>(size);
+            for (Pair<ByteComparable, PrimaryKeys> p : toMerge)
+                if (p != null && p.right != null && !p.right.isEmpty())
+                    keyIterators.add(p.right.iterator());
+
+            Iterator<PrimaryKey> primaryKeys = MergeIterator.get(keyIterators, comparator, new IdentityReducer<>());
+            return Pair.create(term, primaryKeys);
+        }
+
+        @Override
+        public void onKeyChange()
+        {
+            for (int i = 0; i < size; i++)
+                toMerge[i] = null;
+        }
+
+        private ByteComparable getTerm()
+        {
+            for (Pair<ByteComparable, PrimaryKeys> p : toMerge)
+            {
+                if (p != null)
+                    return p.left;
+            }
+            throw new IllegalStateException("Term must exist in IndexMemtable.");
+        }
+
+        class IdentityReducer<In> extends MergeIterator.Reducer<In, In>
```

Comment:
If I understood correctly, a `Reducer` combines multiple values of type `In` and produces one value of type `Out`.
This reducer here simply ignores all the previous values and returns the last value added to the set of values to reduce. If we give it different objects to reduce, it would lose data, so that doesn't make much sense. I can only see it being used for duplicate removal...

Why do we need a thing like this? I guess we're not really reducing anything here, because we're using it to merge multiple lists of primary keys. Because of the way the `MergeIterator` works, it will match the same keys together, and  this `IdentityReducer` will be given identical keys to be reduced. Effectively, this identity reducer will make sure the duplicates are eliminated. Is it correct? 

However, I guess there shouldn't be any duplicate primary keys, because we're merging lists from different shards, and each shard indexes only its own (disjuntive) range of keys - the same key can't appear in more than one shard, correct?

So I'm a bit confused.
Consider adding an explanation.